optashift-employee-rostering-server/src/main/java/org/optaplanner/openshift/employeerostering/server/roster/RosterRestServiceImpl.java

```diff
@@ -359,4 +376,146 @@ public RosterState getRosterState(Integer tenantId) {
+                .filter((e) -> employees.contains(e.getKey()))
+                .collect(Collectors.toMap((e) -> e.getKey().getId(),
+                        (e) -> e.getValue()));
+    }
+
+    private Map<Shift, IndictmentView> getShiftIndictmentMap(int tenantId) {
+        Solver<Roster> solver = solverManager.getSolver(tenantId);
+        if (null != solver) {
+            Roster roster = solver.getBestSolution();
+            try (ScoreDirector<Roster> scoreDirector = solver.getScoreDirectorFactory()
+                    .buildScoreDirector()) {
+                scoreDirector.setWorkingSolution(roster);
+                Map<Shift, IndictmentView> indictmentMap = new HashMap<>();
+                scoreDirector.getIndictmentMap().forEach((k, v) -> {
+                    if (k instanceof Shift) {
+                        indictmentMap.put((Shift) k, new IndictmentView(v));
+                    }
+                });
+                tenantIdToShiftIndictmentMap.put(tenantId, indictmentMap);
+                tenantIdToShiftIndictmentMapUpdateTime.put(tenantId, OffsetDateTime.now());
+                return indictmentMap;
+            }
+        } else {
+            OffsetDateTime lastUpdateTime = getLastUpdateTime(tenantId);
+            OffsetDateTime lastMapUpdateTime = tenantIdToShiftIndictmentMapUpdateTime.getOrDefault(tenantId, OffsetDateTime.MIN);
+            if (lastUpdateTime.isBefore(lastMapUpdateTime)) {
+                return tenantIdToShiftIndictmentMap.get(tenantId);
+            }
+            throw new OutOfDateException(tenantIdToShiftIndictmentMap.getOrDefault(tenantId, Collections.emptyMap()),
+                    lastMapUpdateTime, lastUpdateTime);
+        }
+    }
+
+    private Map<Employee, IndictmentView> getEmployeeIndictmentMap(int tenantId) {
+        Solver<Roster> solver = solverManager.getSolver(tenantId);
+        if (null != solver) {
+            Roster roster = solver.getBestSolution();
+            try (ScoreDirector<Roster> scoreDirector = solver.getScoreDirectorFactory()
+                    .buildScoreDirector()) {
+                scoreDirector.setWorkingSolution(roster);
+                Map<Employee, IndictmentView> indictmentMap = new HashMap<>();
+                scoreDirector.getIndictmentMap().forEach((k, v) -> {
+                    if (k instanceof Employee) {
+                        indictmentMap.put((Employee) k, new IndictmentView(v));
+                    }
+                });
+                tenantIdToEmployeeIndictmentMap.put(tenantId, indictmentMap);
+                tenantIdToEmployeeIndictmentMapUpdateTime.put(tenantId, OffsetDateTime.now());
+                return indictmentMap;
+            }
+        } else {
+            OffsetDateTime lastUpdateTime = getLastUpdateTime(tenantId);
+            OffsetDateTime lastMapUpdateTime = tenantIdToEmployeeIndictmentMapUpdateTime.getOrDefault(tenantId, OffsetDateTime.MIN);
+            if (lastUpdateTime.isBefore(lastMapUpdateTime)) {
```

Comment:
use the optimistic version number instead.

Reply:
is it guaranteed that ALL future versions are greater than ALL past versions (I am guessing that since when Version is OffsetDateTime, version is a timestamp, it is still a timestamp when it is a long)?

Reply:
I use "[count(id), sum(version)]" as a hash